enhancements/bpfd-on-ocp.md

```diff
@@ -0,0 +1,424 @@
+   RBAC. 
+
+8. As an OCP Developer/Admin I want to limit the container permissions required
+   to load and run BPF programs. 
+
+   In OCP today loading and managing BPF programs often requires that the pod be
+   privileged and have both the CAP_BPF and CAP_NET_ADMIN capabilities. This is
+   ultimately a major security concern and should be mitigated to ensure
+   applications can load BPF programs without such privileges.
+
+### Goals
+
+The goal of this proposal is to provide OCP developers and customers with a
+consistent and more secure experience using BPF. Whether they are consuming BPF
+as a component part of OCP, from a partner product, or producing BPF themselves.
+
+If we succeed in our goal, we will provide a solution, or at least the
+infrastructure for a solution, for the user stories listed above. In particular:
+
+- It will be simpler for OCP developers to use BPF in their work.
+- Red Hat teams can collaborate on BPF development.
+- OCP administrators will be able to easily discover what BPF has been loaded in
+  their cluster, who loaded it, and for what purpose.
+- Security will be improved because applications leveraging BPF will not need to
+  run in privileged containers.
+
+### Non-Goals
+
+We do not want to mandate that customers ONLY use our method of working with BPF
+in OpenShift.
+
+We do not intend to provide any controls for blocking the loading of BPF, but we
+may alert customers when incompatibilities are detected. Furthermore, we would
+like to defer the enforcement of ANY policy to other products (i.e ACS).
+
+## Proposal
+
+As mentioned in the summary, this enhancement proposes the following:
+
+- That the bpfd operator be included in OCP as a foundational operator.
+- That the bpfd operator be used by internal Red Hat OCP applications and
+  components for the - loading and management of required BPF programs.
+- That bpfd be recommended and supported for use by OCP partners and users to
+  load and manage their BPF programs when running on OCP
+
+### Workflow Description
+
+In general bpfd and the bpfd operator are designed to work with applications in
+three different deployment models: 
+
+1. **Manual Deployment** of a `bpfProgramConfig` by a cluster administrator with
+   no accompanying application. 
+2. **Manual Deployment** of a `bpfProgramConfig` by a cluster administrator with
+   an accompanying application. 
+3. **Dynamic Deployment** of a `bpfProgramConfig` where the application deploys
+   and manages the BPF program while also interacting with it on a per-node
+   basis. 
+
+For all deployment scenarios, the application or a portion of it needs to be
+deployed as a per-node agent with a mount-point linked to the bpfd-fs
+`/run/bpfd/fs/`. Each application agent will then use the bpfd-helpers library
+(still WIP) to fetch the node-specific `bpfProgram` which is used internally to
+determine the absolute pin path of the maps on the node.
```

Comment:
Where does the "per-node application agent" come from?  Does the application provide it?  Or that's something the bpf operator will provide+deploy?  I would have assumed the latter, but `the application or a portion of it needs to be
deployed as a per-node agent` makes it sound like it comes from the application.

Reply:
So today the application _would_ need to provide a per node agent in order to program + interact with it's BPF application.

We have looked into providing a "centralize map api" which could be represented via a K8s CRD. However due to the diverse nature of bpf in linux (i.e the numerous disparate Map and Program types), and the fact that bpf map interaction is usually tightly coupled to each individual application, we determined it was not really possible.

Instead bpfd + the bpfd operator focuses on making the UX for generic bits(loading and managing BPF) much better.

Reply:
can we get a full workflow step list from the application's perspective then?  sounds like it's something like:

1) application creates a daemonset to run its node agent
2) application creates bpfProgramConfig resource
3) bpfd operator creates bpfProgram resource from the bpfProgramConfig
4) bpfd node agent(?) consumes bpfProgram resource, uses it to download/install/run the bpfProgram binary
5) application node-agent programs/interacts with running bpfProgram

it's helpful for understanding the personas and permissions involved

Reply:
Hello Ben.  We have a couple of links that will help with this, and I can also update the pr with this info.
1. [High-level description of what the user/app needs to do](https://docs.google.com/presentation/d/1eg1OPKS5ahMTUauPLyOTiKOMrYa8sQugYzMK4NpYCA8/edit#slide=id.g1fb774bd2ea_0_74).
2. [Description of the bpfd workflow for an example application we have](https://github.com/redhat-et/bpfd/blob/main/docs/example-bpf.md#deploying-on-kubernetes).

Your flow is pretty close.  Here's a summary of what happens:

1. application creates a daemonset to run its node agent
2. application creates a BpfProgramConfig object.
3. bpfd-agent (running on each node) consumes the BpfProgramConfig objects. 
4. bpfd-agent creates a BpfProgram object for that node
5. bpfd-agent makes request to bpfd daemon to load or unload the bpf program(s) as needed by making gRPC calls
6. bpfd daemon does the actual loading/unloading and map pinning for the bpfd program.
7. bpfd-agent updates the status of the BpfProgram object (including map pin point(s))
8. application interacts with BpfProgram object to get map location
9. application interacts with bpf program via the maps that are created

Separately, the bpfd-operator watches all BpfProgram objects, and updates the status of the BpfProgramConfig object indicating whether the BPF program has been applied to all the desired nodes.

Note that without bpfd, bpf-based applications still need to do the things that bpfd is doing, it's just that they each need to "roll their own", and they don't get the benefits of using bpfd that we've outlined.

Reply:
I'll also add that this workflow describes the "Dynamic Deployment of a bpfProgramConfig" model mentioned in the enhancement request.

Reply:
I think these examples could be done in terms of very concrete examples. I would find that a bit easier to follow than a more generic example.

* Assume we replace our NetworkPolicy implementation using eBPF instead of OVS flows. What does that workflow look like from cluster-network-operator and/or ovn-kubernetes? This example demonstrates how this has to work to support a core network feature (not an add-on).

* Look at the network observability operator and its use of eBPF. How would that operator be adjusted to use bpfd instead? This will demonstrate the workflow from the perspective of an add-on component.

Reply:
> Your flow is pretty close. Here's a summary of what happens:

thanks, that helped

Reply:
I agree, some concrete examples would be useful.

In this comment thread, is "application" an operator? Or a BPF construct? Or a generic term for some combination of those? On the one hand it sounds like the application runs centrally, because it's creating a DaemonSet and BpfProgramConfig object (steps 1 & 2), but then it also sounds like it is running on (or at least looking at) each Node because it is interacting with the unique BpfProgram for each Node (steps 8 & 9). Using a different name for the central thing and the per-node thing might help clarify the text.